src/extent.c

```diff
@@ -2045,13 +2039,16 @@ extent_purge_forced_wrapper(tsdn_t *tsdn, arena_t *arena,
 	    offset, length, false);
 }
 
-#ifdef JEMALLOC_MAPS_COALESCE
 static bool
 extent_split_default(extent_hooks_t *extent_hooks, void *addr, size_t size,
     size_t size_a, size_t size_b, bool committed, unsigned arena_ind) {
-	return !maps_coalesce;
+	if (maps_coalesce) {
+		return false;
+	}
+
+	/* See comments in extent_merge_default_impl(). */
+	return opt_retain ? false : true;
```

Comment:
Why does extent splitting return an error only if retain is off? I see why we might want to treat this differently, but erroring out on splitting seems like a weird way to do so.

Also, I think this whole thing is most clearly written as: `return !(maps_coalesce || opt_retain)`.

Reply:
Will refactor this. I wanted to separate the Windows logic and add some comments. Let me make that happen.

I think we want to consider the returning true as "not allowed to split" instead of a real error; without retain, purging does MEM_RELEASE which only accepts the whole region. In that case we don't allow split / merge. The user custom hooks might return true as a way to say not-allowed as well, e.g. in hhvm for arenas owning reserved huge pages (but with regular pages as back up), they may check if an extent is part of the huge pages and disallow purging.